File: examples/receipt-scanner/python/src/store.py
import json
from datetime import datetime, timezone
from pathlib import Path

from .contract import ReceiptScanResult

DB_PATH = Path(__file__).parent.parent / "scans.json"


def _load() -> list[dict]:
    if not DB_PATH.exists():
        return []
    return json.loads(DB_PATH.read_text())


def _save(scans: list[dict]) -> None:
    DB_PATH.write_text(json.dumps(scans, indent=2))


def add_scan(scan: ReceiptScanResult, file: str) -> dict:
    scans = _load()
    record = {
        **scan.model_dump(),
        "id": len(scans) + 1,
        "scanned_at": datetime.now(timezone.utc).isoformat(),
        "file": file,
    }
    print(
        f"[Boundary] Saving scan #{record['id']}: {record['vendor']}"
        f" — {record['currency']} {record['amount']:.2f} ({record['category']})"
    )
    scans.append(record)
    _save(scans)
    print(
        f"[Boundary] Scan #{record['id']} written to store"
        f" (total: {len(scans)} record(s))"
    )
    return record


def list_scans() -> list[dict]:
    return _load()

File: examples/receipt-scanner/node/src/store.ts
import fs from "fs";
import path from "path";
import type { Receipt } from "./contract.js";

const DB_PATH = path.resolve("expenses.json");

export interface Expense extends Receipt {
  id: number;
  scannedAt: string;
  file: string;
}

function load(): Expense[] {
  if (!fs.existsSync(DB_PATH)) return [];
  return JSON.parse(fs.readFileSync(DB_PATH, "utf-8"));
}

function save(expenses: Expense[]): void {
  fs.writeFileSync(DB_PATH, JSON.stringify(expenses, null, 2));
}

export function addExpense(receipt: Receipt, file: string): Expense {
  const expenses = load();
  const expense: Expense = {
    ...receipt,
    id: expenses.length + 1,
    scannedAt: new Date().toISOString(),
    file,
  };
  console.log(`[Boundary] Saving expense #${expense.id}: ${expense.vendor} — ${expense.currency} ${expense.amount.toFixed(2)} (${expense.category})`);
  expenses.push(expense);
  save(expenses);
  console.log(`[Boundary] Expense #${expense.id} written to store (total: ${expenses.length} record(s))`);
  return expense;
}

export function listExpenses(): Expense[] {
  return load();
}
